### PR TITLE
add openshift/installer 'bazel build tarball' test to prow

### PR DESCRIPTION
**Pull request by sallyom**
@bbguimaraes another installer job for prow, thanks

### CODE REVIEW

File: ci-operator/jobs/openshift/installer/openshift-installer-presubmits.yaml [hunk 1]
```diff
@@ -153,3 +153,22 @@ presubmits:
         - sh
         args:
         - ./hack/tf-fmt.sh
+
+  - name: ci-pull-openshift-installer-bazel-build-tarball
+    trigger: "(?m)^/test build-tarball"
+    rerun_command: "/test build-tarball"
+    context: ci/prow/build-tarball
+    always_run: true
+    skip_report: false
```
Comment (bbguimaraes) [inline ci-operator/jobs/openshift/installer/openshift-installer-presubmits.yaml]: That's the default value, so it's not really necessary.

**Comment by wking** [comment]
This job is [failing][1] [with][2]:

```
+ bazel --output_base=/tmp build tarball
Error: $USER is not set, and unable to look up name of current user: (error: 0): Success
```

I'm looking into it, but thought I'd post to this PR in case someone more familiar with the Prow execution environment understood what was going on without digging ;).

[1]: https://github.com/openshift/installer/pull/123#issuecomment-412952622
[2]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/123/ci-pull-openshift-installer-bazel-build-tarball/1/build-log.txt

**Comment by wking** [comment]
The error message is coming from [this block][1], where Bazel is finding an unset `$USER`, falling back on `getpwuid(getuid())` and getting either a `NULL` struct pointer or `pw_name`.  I'm not sure how that would happen in Prow but not Travis, because in both cases Bazel is running inside a container.  Maybe Prow has stricter seccomp settings or some such?

[1]: https://github.com/bazelbuild/bazel/blob/0.16.1/src/main/cpp/blaze_util_posix.cc#L654-L664

**Comment by bbguimaraes** [comment]
@wking: from the man page:

```
The getpwnam() and getpwuid() functions return a pointer to a passwd structure,
or NULL if the matching entry is not found or an error occurs.  If an error
occurs, errno is set appropriately.
```

It means there's no entry for that uid in the passwd file, which makes sense
because containers in Openshift execute with arbitrary uids.

That seems like a strange requirement for a build system (but Bazel never
ceases to amaze me).

**Comment by wking** [comment]
It looks like the container is explicitly setting a user:

```console
# podman pull quay.io/coreos/tectonic-builder:bazel-v0.3
# podman history quay.io/coreos/tectonic-builder:bazel-v0.3 | grep USER
<missing>      5 months ago   /bin/sh -c #(nop) USER [bazel]                  0B
```

I'm not quite sure where that version of the image comes from, but I think the associated `Dockerfile` line is [here][1].  The [`USER` Dockerfile command][2] does not result in the `USER` environment variable being set:

```console
# podman run --rm -it quay.io/coreos/tectonic-builder:bazel-v0.3 id
uid=1000(bazel) gid=1000(bazel) groups=1000(bazel)
# podman run --rm -it quay.io/coreos/tectonic-builder:bazel-v0.3 env | grep USER
```


[1]: https://github.com/openshift/installer/blob/5ac7cfb476972284e7dd25bc14d8d82442f439bc/images/tectonic-builder/Dockerfile.bazel#L27
[2]: https://docs.docker.com/engine/reference/builder/#user
[2]: https://github.com/bazelbuild/bazel/blob/0.16.1/src/main/cpp/blaze_util_posix.cc#L655

**Comment by wking** [comment]
> That is ignored when the pod is assigned a uid by the security policy in Openshift...

Why would OpenShift care about the container-side UID (I can understand caring about the host-side UID)?

> I'm not familiar with Bazel, but you can always set the USER variable yourself in the job/script if its value is irrelevant.

Grepping, it looks like the value may only be used for [an output directory name][1].  We're [explicitly setting the output directory][2], and we aren't installing anything, so I expect we are safe to explicitly set `USER`.  I'll work up a PR.

[1]: https://github.com/bazelbuild/bazel/blob/0.16.1/src/main/cpp/startup_options.cc#L120-L121
[2]: https://github.com/openshift/installer/blob/5ac7cfb476972284e7dd25bc14d8d82442f439bc/hack/test-bazel-build-tarball.sh#L4

**Comment by wking** [comment]
> ... so I expect we are safe to explicitly set `USER`...

Once we have a fix, I expect we'll want to update the Prow job config in this repo.  But for testing fixes, I'll use local adjustments in the hack script itself.  Setting `USER` did get us past the [earlier error][1].  Now we're dying [with][2]:

```
+ USER=bazel-testing bazel --output_base=/tmp build tarball
Error: mkdir('/.cache/bazel/_bazel_bazel-testing'): (error: 13): Permission denied
```

I'll keep adding command-line arguments and environment variables until I get the script to pass...

[1]: https://github.com/openshift/release/pull/1178#issuecomment-412953310
[2]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/123/ci-pull-openshift-installer-bazel-build-tarball/3/build-log.txt

**Comment by wking** [comment]
Setting both `USER` and `HOME` was sufficient to get Bazel up off the ground.  But then a Python invocation crashed [with][1]:

```
ERROR: /home/prow/go/src/github.com/openshift/installer/BUILD.bazel:106:1: error executing shell command: 'bazel-out/host/bin/external/bazel_tools/tools/build_defs/pkg/build_tar --flagfile=bazel-out/k8-fastbuild/bin/tf_bin.args' failed (Exit 1)
Traceback (most recent call last):
  File "/usr/lib/python2.7/site.py", line 554, in <module>
    main()
  File "/usr/lib/python2.7/site.py", line 536, in main
    known_paths = addusersitepackages(known_paths)
  File "/usr/lib/python2.7/site.py", line 272, in addusersitepackages
    user_site = getusersitepackages()
  File "/usr/lib/python2.7/site.py", line 247, in getusersitepackages
    user_base = getuserbase() # this will also set USER_BASE
  File "/usr/lib/python2.7/site.py", line 237, in getuserbase
    USER_BASE = get_config_var('userbase')
  File "/usr/lib/python2.7/sysconfig.py", line 582, in get_config_var
    return get_config_vars().get(name)
  File "/usr/lib/python2.7/sysconfig.py", line 533, in get_config_vars
    _CONFIG_VARS['userbase'] = _getuserbase()
  File "/usr/lib/python2.7/sysconfig.py", line 210, in _getuserbase
    return env_base if env_base else joinuser("~", ".local")
  File "/usr/lib/python2.7/sysconfig.py", line 196, in joinuser
    return os.path.expanduser(os.path.join(*args))
  File "/usr/lib/python2.7/posixpath.py", line 262, in expanduser
    userhome = pwd.getpwuid(os.getuid()).pw_dir
KeyError: 'getpwuid(): uid not found: 1000130000'
```

Python [should be looking at `$HOME` first as well][2], so I'm not sure yet why it's falling back to the `getpwuid` call.  Maybe we need to do something to get that environment variable through when Bazel drops to Python.

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/123/ci-pull-openshift-installer-bazel-build-tarball/4/build-log.txt
[2]: https://github.com/python/cpython/blob/1f34aece28d143edb94ca202e661364ca394dc8c/Lib/posixpath.py#L260-L262

**Comment by stevekuznetsov** [comment]
@smarterclayton can comment on security implications of container-side UUID and why OpenShift cares

@bentheelder might have some comments on running `bazel` in containers for testing -- much of upstream k8s testing does that, but I am not familiar with it

**Comment by wking** [comment]
> Maybe we need to do something to get that environment variable through when Bazel drops to Python.

Bazel's `build_tar` [invocation][1] sets [`use_default_shell_env=True`][2].  That seems (at least from the [design docs][3]) to mean that only environment variables declared with `--action_env` are passed through to the action executable.  I'll update the script to set `HOME` that way as well.

[1]: https://github.com/bazelbuild/bazel/blob/0.16.1/tools/build_defs/pkg/pkg.bzl#L82-L89
[2]: https://github.com/bazelbuild/bazel/blob/0.16.1/tools/build_defs/pkg/pkg.bzl#L88
[3]: https://bazel.build/designs/2016/06/21/environment.html

**Comment by BenTheElder** [comment]
k/k prow jobs do typically set `$HOME` and `$USER`. For [local bazel containers](https://github.com/kubernetes/test-infra/tree/master/planter) I've also found a fake passwd entry to be useful for eg python and git.

**Comment by wking** [comment]
> @smarterclayton can comment on security implications of container-side UUID and why OpenShift cares

Ah, I think I found the reason [over here][1]:

> By default, all containers that we try and launch within OpenShift, are set blocked from “RunAsAny” which basically means that they are not allowed to use a root user within the container. This prevents root actions such as chown or chmod from being run and is a sensible security precaution as, should a user be able to perform a local exploit to break out of the container, then they would not be running as root on the underlying container host. NB what about user-namespaces some of you are no doubt asking, these are definitely coming but the testing/hardening process is taking a while and whilst companies such as Red Hat are working hard in this space, there is still a way to go until they are ready for the mainstream.

That blog post is from 2016; so I'm not sure the user-namespace mainstream-ness caveat still applies.  But that's probably still why OpenShift is doing this.

[1]: https://blog.openshift.com/getting-any-docker-image-running-in-your-own-openshift-cluster/

**Comment by BenTheElder** [comment]
You should be able to run without privileged containers if you disabling the sandboxing IIRC, something like `--spawn_strategy=local` either as an extra flag to the build or in a .bazelrc as `build --spawn_strategy=local`. We don't do this because the sandboxing is important to ensuring correctness of known inputs for each step, without enforcing that we may cache incorrectly; and we use cluster-level isolation where we are concerned.

**Pull request merged**